src/System.Private.CoreLib/shared/System/Reflection/AssemblyName.cs

```diff
@@ -0,0 +1,602 @@
+        /// match the intent of this api, this api has been broken this way since its debut and we cannot
+        /// change its behavior now.
+        /// </summary>
+        public static bool ReferenceMatchesDefinition(AssemblyName reference, AssemblyName definition)
+        {
+            if (object.ReferenceEquals(reference, definition))
+                return true;
+
+            if (reference == null)
+                throw new ArgumentNullException(nameof(reference));
+
+            if (definition == null)
+                throw new ArgumentNullException(nameof(definition));
+
+            string refName = reference.Name ?? string.Empty;
+            string defName = definition.Name ?? string.Empty;
+            return refName.Equals(defName, StringComparison.OrdinalIgnoreCase);
+        }
+
+        internal void SetProcArchIndex(PortableExecutableKinds pek, ImageFileMachine ifm)
+        {
+            ProcessorArchitecture = CalculateProcArchIndex(pek, ifm, _flags);
+        }
+
+        internal static ProcessorArchitecture CalculateProcArchIndex(PortableExecutableKinds pek, ImageFileMachine ifm, AssemblyNameFlags flags)
+        {
+            if (((uint)flags & 0xF0) == 0x70)
+                return ProcessorArchitecture.None;
+
+            if ((pek & System.Reflection.PortableExecutableKinds.PE32Plus) == System.Reflection.PortableExecutableKinds.PE32Plus)
+            {
+                switch (ifm)
+                {
+                    case System.Reflection.ImageFileMachine.IA64:
+                        return ProcessorArchitecture.IA64;
+                    case System.Reflection.ImageFileMachine.AMD64:
+                        return ProcessorArchitecture.Amd64;
+                    case System.Reflection.ImageFileMachine.I386:
+                        if ((pek & System.Reflection.PortableExecutableKinds.ILOnly) == System.Reflection.PortableExecutableKinds.ILOnly)
+                            return ProcessorArchitecture.MSIL;
+                        break;
+                }
+            }
+            else
+            {
+                if (ifm == System.Reflection.ImageFileMachine.I386)
+                {
+                    if ((pek & System.Reflection.PortableExecutableKinds.Required32Bit) == System.Reflection.PortableExecutableKinds.Required32Bit)
+                        return ProcessorArchitecture.X86;
+
+                    if ((pek & System.Reflection.PortableExecutableKinds.ILOnly) == System.Reflection.PortableExecutableKinds.ILOnly)
+                        return ProcessorArchitecture.MSIL;
+
+                    return ProcessorArchitecture.X86;
+                }
+                if (ifm == System.Reflection.ImageFileMachine.ARM)
+                {
+                    return ProcessorArchitecture.Arm;
+                }
+            }
+            return ProcessorArchitecture.None;
+        }
+
+        internal void Init(string name,
```

Comment:
I would keep Clone done the same way as it is in CoreRT, and change this method into CoreCLR specific internal constructor. Or even delete it - the cloning done by this method is unnecessary when it is called by VM; the VM can just set the fields directly.